File: src/vibey/infrastructure/hub/lanes.py
# Made with ❤️ by [Vibey](https://the-vibey-project.github.io/vibey/), Developed by [Adam Matthew Steinberger](https://vibewithadam.matthewsteinberger.com/) ([@adammatthewsteinberger](https://github.com/adammatthewsteinberger/)).
"""Every lane on this computer, found where the loops write them.

A lane is one run of a `*loop` runner. Each writes `<cwd>/<state_dir>/runs/<run id>/
events.jsonl` (the descriptors' run-events template), and this finds them the way the VS
Code extension's `LaneTracker` does: in each root, and in every directory directly inside
it. A lane's position is its file's size in bytes -- the offset a reader resumes after
(sub-doctrine 10.g) -- never a timestamp.

What it reports is what the files prove and nothing more (10.f): `finished` when the
runner's last snapshot says the run ended, `running` when the file changed within
`quiet_after` seconds, `quiet` when it has not; a lane whose file is older than
`recent_within` is not listed. The transcript itself -- turns, tools, tokens -- is the
live feed's to tail (`vibey serve`'s lane stream), not this list's.
"""

import json
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Final

QUIET_AFTER: Final = 120.0
"""Seconds without an event after which a running lane reads as quiet."""

RECENT_WITHIN: Final = 24 * 60 * 60.0
"""A lane whose last event is older than this many seconds is not listed."""

ENDED: Final[dict[str, str]] = {
    "completed": "completed",
    "failed": "failed",
    "winding_down": "stopped",
}
"""A runner snapshot's status, read as how the lane ended."""


@dataclass(frozen=True, slots=True)
class LaneEngine:
    """An engine a lane may belong to, and the directory its runner writes under."""

    engine_id: str
    state_dir: str


class LaneScanner:
    """Lists the lanes under `roots`.

    Declared by `interfaces/lanes_interface.py::LaneScannerInterface`."""

    def __init__(
        self,
        *,
        engines: Sequence[LaneEngine],
        roots: Sequence[Path],
        now: Callable[[], float],
        quiet_after: float = QUIET_AFTER,
        recent_within: float = RECENT_WITHIN,
    ) -> None:
        self._engines = tuple(engines)
        self._roots = tuple(roots)
        self._now = now
        self._quiet_after = quiet_after
        self._recent_within = recent_within

    def lanes(self) -> list[dict[str, object]]:
        """Every recent lane, newest first."""
        now = self._now()
        found: dict[Path, dict[str, object]] = {}
        for events, engine, cwd, run_id in self._discover():
            if events in found:
                continue
            stat = events.stat()
            if now - stat.st_mtime > self._recent_within:
                continue
            outcome = self._outcome(events.parent)
            state = (
                "finished"
                if outcome is not None
                else "quiet"
                if now - stat.st_mtime > self._quiet_after
                else "running"
            )
            found[events] = {
                "id": run_id,
                "engine": engine.engine_id,
                "cwd": str(cwd),
                "events_path": str(events),
                "label": f"{cwd.name} · {engine.engine_id}",
                "state": state,
                "outcome": outcome,
                "offset": stat.st_size,
                "last_event_at": stat.st_mtime,
            }
        return sorted(found.values(), key=self._last_event, reverse=True)

    def _discover(self) -> list[tuple[Path, LaneEngine, Path, str]]:
        found: list[tuple[Path, LaneEngine, Path, str]] = []
        for root in self._roots:
            for cwd in (root, *self._children(root)):
                for engine in self._engines:
                    for run in self._children(cwd / engine.state_dir / "runs"):
                        events = run / "events.jsonl"
                        if events.is_file():
                            found.append((events, engine, cwd, run.name))
        return found

    @staticmethod
    def _last_event(lane: dict[str, object]) -> float:
        value = lane["last_event_at"]
        return value if isinstance(value, float) else 0.0

    @staticmethod
    def _children(directory: Path) -> list[Path]:
        try:
            return sorted(
                entry
                for entry in directory.iterdir()
                # A symlink is never walked: it could lead the scan out of its root.
                if entry.is_dir() and not entry.is_symlink() and not entry.name.startswith(".")
            )
        except OSError:
            return []

    @staticmethod
    def _outcome(run: Path) -> str | None:
        try:
            snapshot = json.loads((run / "snapshots" / "latest.json").read_text())
        except (OSError, ValueError):
            return None
        status = snapshot.get("status") if isinstance(snapshot, dict) else None
        return ENDED.get(status) if isinstance(status, str) else None

File: clients/vscode/src/core/lanes.ts
// Made with ❤️ by [Vibey](https://the-vibey-project.github.io/vibey/), Developed by [Adam Matthew Steinberger](https://vibewithadam.matthewsteinberger.com/) ([@adammatthewsteinberger](https://github.com/adammatthewsteinberger/)).
/**
 * Every lane running on this computer, found where the loops write them: `<state_dir>/runs/
 * <run id>/events.jsonl` in each worktree under the storm home, in the storm home itself,
 * and in the open folders. Lanes the editor started are among them, and so are a storm's
 * and a terminal's.
 *
 * Each lane is read by byte offset (sub-doctrine 10.g) and folded into what a person wants
 * at a glance: running, quiet or finished, its turn, the tool it is running, its tokens and
 * how long it has run. A lane with no event for a while is quiet; nothing here ever stops
 * one. The local runner's final snapshot (gptossloop's or qwenloop's) says how a lane ended
 * when its events do not. Declared by `interfaces/lanes-interface.ts`.
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import type { JsonlTailInterface } from './interfaces/jsonl-interface';
import type { Lane, LaneEngine, LaneTrackerInterface } from './interfaces/lanes-interface';
import type { RunItem } from './interfaces/run-events-interface';
import { RunTranscript } from './run-events';

interface Followed {
  readonly engine: LaneEngine;
  readonly cwd: string;
  readonly id: string;
  readonly transcript: RunTranscript;
  offset: number;
  startedAt: number;
  lastEventAt: number;
  tool?: string;
  outcome?: string;
}

export class LaneTracker implements LaneTrackerInterface {
  private readonly followed = new Map<string, Followed>();
  private current: readonly Lane[] = [];

  constructor(
    private readonly engines: readonly LaneEngine[],
    /** Directories to look in: each is searched, and so is every directory directly inside it. */
    private readonly roots: () => readonly string[],
    private readonly tail: JsonlTailInterface,
    private readonly now: () => number,
    /** Silence longer than this makes a running lane quiet. */
    private readonly quietMs: number,
    /** A lane whose last event is older than this is no longer listed. */
    private readonly recentMs: number,
  ) {}

  lanes(): readonly Lane[] {
    return this.current;
  }

  items(eventsPath: string): readonly RunItem[] {
    return this.followed.get(eventsPath)?.transcript.items() ?? [];
  }

  refresh(): readonly Lane[] {
    for (const [eventsPath, engine, cwd, id] of this.discover()) {
      if (!this.followed.has(eventsPath)) {
        this.followed.set(eventsPath, {
          engine,
          cwd,
          id,
          transcript: new RunTranscript(),
          offset: 0,
          startedAt: LaneTracker.born(fs.statSync(eventsPath)),
          lastEventAt: 0,
        });
      }
    }
    const now = this.now();
    const lanes: Lane[] = [];
    for (const [eventsPath, lane] of this.followed) {
      let modified: number;
      try {
        modified = fs.statSync(eventsPath).mtimeMs;
      } catch {
        this.followed.delete(eventsPath);
        continue;
      }
      lane.lastEventAt = modified;
      this.read(eventsPath, lane);
      if (now - lane.lastEventAt > this.recentMs) {
        continue;
      }
      const finished = lane.outcome !== undefined;
      lanes.push({
        id: lane.id,
        engine: lane.engine.engineId,
        cwd: lane.cwd,
        eventsPath,
        label: `${path.basename(lane.cwd)} · ${lane.engine.engineId}`,
        state: finished ? 'finished' : now - lane.lastEventAt > this.quietMs ? 'quiet' : 'running',
        ...(lane.outcome === undefined ? {} : { outcome: lane.outcome }),
        turn: lane.transcript.turns,
        ...(lane.tool === undefined || finished ? {} : { tool: lane.tool }),
        startedAt: lane.startedAt,
        lastEventAt: lane.lastEventAt,
        inputTokens: lane.transcript.inputTokens,
        outputTokens: lane.transcript.outputTokens,
      });
    }
    this.current = lanes.sort((left, right) => right.startedAt - left.startedAt);
    return this.current;
  }

  private read(eventsPath: string, lane: Followed): void {
    const chunk = this.tail.read(eventsPath, lane.offset);
    lane.offset = chunk.nextOffset;
    for (const event of chunk.records) {
      for (const patch of lane.transcript.accept(event, lane.engine.envelope)) {
        if (patch.op === 'add' && patch.item.kind === 'tool-call') {
          lane.tool = patch.item.name;
        } else if (patch.op === 'add' && patch.item.kind === 'tool-result') {
          lane.tool = undefined;
        }
      }
    }
    if (lane.transcript.completed) {
      lane.outcome = 'completed';
    } else if (lane.transcript.failure !== undefined) {
      lane.outcome = 'failed';
    }
    lane.outcome ??= LaneTracker.snapshot(path.dirname(eventsPath));
  }

  /** Every events file under the roots, with its engine, worktree and run id. */
  private discover(): [string, LaneEngine, string, string][] {
    const found: [string, LaneEngine, string, string][] = [];
    const seen = new Set<string>();
    for (const root of this.roots()) {
      for (const cwd of [root, ...LaneTracker.children(root)]) {
        for (const engine of this.engines) {
          const runs = path.join(cwd, engine.stateDir, 'runs');
          for (const id of LaneTracker.children(runs).map((directory) => path.basename(directory))) {
            const eventsPath = path.join(runs, id, 'events.jsonl');
            if (!seen.has(eventsPath) && fs.existsSync(eventsPath)) {
              seen.add(eventsPath);
              found.push([eventsPath, engine, cwd, id]);
            }
          }
        }
      }
    }
    return found;
  }

  /** The directories directly inside `directory`; none when it cannot be read. */
  private static children(directory: string): string[] {
    try {
      return fs
        .readdirSync(directory, { withFileTypes: true })
        .filter((entry) => entry.isDirectory() && !entry.name.startsWith('.'))
        .map((entry) => path.join(directory, entry.name));
    } catch {
      return [];
    }
  }

  /** The local runner writes its last snapshot when a run ends; its status says how. */
  private static snapshot(runDirectory: string): string | undefined {
    try {
      const snapshot: unknown = JSON.parse(fs.readFileSync(path.join(runDirectory, 'snapshots', 'latest.json'), 'utf8'));
      const status = (snapshot as { status?: unknown }).status;
      return status === 'completed' || status === 'failed'
        ? status
        : status === 'winding_down'
          ? 'stopped'
          : undefined;
    } catch {
      return undefined;
    }
  }

  /** When the events file was made, the lane's start: its birth time where the filesystem keeps one. */
  static born(stat: { readonly birthtimeMs: number; readonly ctimeMs: number }): number {
    return stat.birthtimeMs > 0 ? stat.birthtimeMs : stat.ctimeMs;
  }
}
